momento-sdk/src/main/java/momento/sdk/ScsDataGrpcStubsManager.java

```diff
@@ -25,22 +25,65 @@ final class ScsDataGrpcStubsManager implements Closeable {
   private final ManagedChannel channel;
   private final ScsGrpc.ScsFutureStub futureStub;
   private final Duration deadline;
+  private final ScheduledExecutorService retryScheduler;
+  private final ExecutorService retryExecutor;
+
+  // An arbitary selection of twice the number of available processors.
+  private static final int MAX_RETRY_THREAD_POOL_SIZE =
+      Runtime.getRuntime().availableProcessors() * 2;
+  // Timeout to keep threads idle/alive in the retry thread pool
+  private static final long RETRY_THREAD_POOL_KEEP_ALIVE_SECONDS = 60L;
 
   ScsDataGrpcStubsManager(
       @Nonnull CredentialProvider credentialProvider, @Nonnull Configuration configuration) {
     this.deadline = configuration.getTransportStrategy().getGrpcConfiguration().getDeadline();
 
-    this.channel = setupChannel(credentialProvider);
+    /**
+     * These two executors are used by {@link RetryClientInterceptor} to schedule and execute
+     * retries on failed gRPC requests. One is a single threaded scheduling executor {@link
+     * ScheduledExecutorService} that's only responsible to schedule retries with a given delay. The
+     * other executor {@link ThreadPoolExecutor} is a dynamic executor that spawns threads as
+     * necessary and executes the retries (essentially doing the I/O work).
+     *
+     * <p>{@link ScheduledExecutorService} creates idle threads and hence the choice of using two
+     * executors. The small overhead of a single thread doing the scheduling negates the
+     * implications of creating potentially hundreds of idle threads with the
+     * ScheduledExecutorService. This executor is chosen to avoid a Thread.sleep() while doing
+     * retries. There are no available configurations to ScheduledExecutorService such that it grows
+     * dynamically.
+     *
+     * <p>The {@link ThreadPoolExecutor} has a size of 0 and grows onDemand. The maximumPoolSize is
+     * present to cap the number of threads we create for retries; where the {@link
+     * LinkedBlockingQueue} essentially stores any pending retries. Note that the keep alive time
+     * for each thread is 60 seconds, beyond which the JVM will destroy the thread. This is the
+     * default value of {@link Executors.newCachedThreadPool()} and also applies in our situation as
+     * even with backoffs, one retry should be below 60 seconds. We aren't directly using a cached
+     * thread pool because it grows unbounded.
+     */
```

Comment:
What would you expect the difference in behavior to be if we just created one executor (the one you are currently calling `retryExecutor`) and gave it a minimum size of 1, and used it for both tasks?

Reply:
The ```retryExecutor``` is a ```ThreadPoolExecutor``` and doesn't provide scheduling semantics. It runs tasks on-demand. The ```retryScheduler``` is a ```ScheduledThreadPoolExecutor``` that schedules tasks with a delay. We could use a ```ScheduledThreadPoolExecutor``` to do both, but it creates idle threads and not on-demand. If we used only this for execution with a size of 1, then all retries will happen sequentially.

Reply:
does the `retryScheduler` sleep in order to achieve the delay? or does it just immediately schedule the task on another thread and add in a sleep to that task that it is scheduling?

Just want to make sure that:
(a) the retryScheduler isn't going to be blocked while we are waiting for the delay to transpire (because if it was then we could only schedule one retry at a time), and
(b) it is worth the (minor) extra complexity of having two thread pools to reason about instead of one.

Reply:
The ```retryScheduler``` does not use sleep internally. Instead, it relies on a combination of a task queue, a thread pool, and a DelayQueue (a priority queue of delayed tasks) to handle scheduling and execution of tasks. I have actually implemented this executor from scratch around a couple years back, we can discuss this more if you'd like in our 1:1 :)

Reply:
The scheduling does happen on a single thread though, but in this case the scheduler is just submitting the task to another executor and moving onto the next Runnable to schedule. It doesn't block but wakes up when there's a task in in's internal queue, and then checks if the delay has elapsed (using epoch time). This check might happen for a bunch of retries in parallel, but again, it's not doing the real work. We can always increase this number though I believe it wouldn't be necessary; I can do a small toy benchmarking if you'd like.

Reply:
it's an implementation detail, we can change it in the future, I trust your judgment if you are confident in the semantics of it and if you believe this is the simplest way to implement it.